GDScript: Clarified/fixed inaccuracies in the built-in function docs.

Meriipu opened the pull request:
note: I have not changed any of the visualscript or mono-stuff, those
might be necessary to change to match the changed parameter-name.

The input to smoothstep is not actually a weight, and the decscription
of smoothstep was pretty hard to understand and easy to misinterpret.

Clarified what it means to be approximately equal.

nearest_po2 does not do what the descriptions says it does. For one,
it returns the same power if the input is a power of 2. Second, it
returns 0 if the input is negative or 0, while the smallest possible
integral power of 2 actually is 1 (2^0 = 1). Due to the implementation
and how it is used in a lot of places, it does not seem wise to change
such a core function however, and I decided it is better to alter the
description of the built-in.

Added a few examples/clarifications/edge-cases.

Meriipu commented:
```
func do():
	printt(-10, nearest_po2(-10))
	printt(-1, nearest_po2(-1))
	for i in range(17):
		printt(i,nearest_po2(i))
```
prints out
```
-10	0
-1	0
0	0
1	1
2	2
3	4
4	4
5	8
6	8
7	8
8	8
9	16
...
```

Meriipu commented:
are these changes OK?

aaronfranke commented on modules/gdscript/gdscript_functions.cpp:
```diff
@@ -1632,7 +1632,7 @@ MethodInfo GDScriptFunctions::get_info(Function p_func) {
 		} break;
 		case MATH_SMOOTHSTEP: {
-			MethodInfo mi("smoothstep", PropertyInfo(Variant::FLOAT, "from"), PropertyInfo(Variant::FLOAT, "to"), PropertyInfo(Variant::FLOAT, "weight"));
+			MethodInfo mi("smoothstep", PropertyInfo(Variant::FLOAT, "from"), PropertyInfo(Variant::FLOAT, "to"), PropertyInfo(Variant::FLOAT, "x"));
```
You're correct that weight is inaccurate, but also "x" isn't very descriptive here. Perhaps "between", since meaningful values exist on the range of "from" to "to" (aka: bewteen them)?

Meriipu replied:
It does not need to be between the two values, though. Values to the right and to the left are meaningful as well, even if they are clamped to the endpoint.

Meriipu replied:
Perhaps `value` or `s`, which while not too descriptive either is consistent with the other mathematical functions like `exp` or `sqrt` (smoothstep is a mathematical function too, after all).

aaronfranke replied:
Well, those valures are valid in that you still get expected outputs, but it's not what I would describe as "meaningful".

Both of your proposals are better than "x" IMO. I'll let you make the final call, but I think "between" still makes sense.

Meriipu replied:
I would say it makes enough sense to have values outside (or just outside) the range too that being overly prescriptive with the naming might be inaccurate.
Another part of why I prefer `x` or `s` is that various other sources do this rather than naming it (wikipedia, opengl https://www.khronos.org/registry/OpenGL-Refpages/gl4/html/smoothstep.xhtml, others, and I do not believe one would find the variable named between in these other sources).
I am leaning towards `between` potentially causing more confusion than it helps, for example since one often says that a value lies between other values, or that one interpolates between other values, and that it is not immediately clear that `between` refers to that the "betweened" value.

aaronfranke commented on modules/gdscript/doc_classes/@GDScript.xml:
```diff
@@ -317,7 +317,7 @@
 			<description>
 				The natural exponential function. It raises the mathematical constant [b]e[/b] to the power of [code]s[/code] and returns it.
-				[b]e[/b] has an approximate value of 2.71828.
+				[b]e[/b] has an approximate value of 2.71828, and can be obtained with [code]exp(1)[/code].
```
Should be something like ", and is equivalent to [code]exp(1)[/code]". We don't want to seem as if we're encouraging actually writing `exp(1)`.

Meriipu replied:
Are there other ways to get the constant e in codot?

aaronfranke replied:
Oh, never mind, I thought we had `e` as an exposed constant, but it's only exposed in C#. Keep this as-is.

Meriipu committed:
GDScript: Clarified/fixed inaccuracies in the built-in function docs.

The input to smoothstep is not actually a weight, and the decscription
of smoothstep was pretty hard to understand and easy to misinterpret.

Clarified what it means to be approximately equal.

nearest_po2 does not do what the descriptions says it does. For one,
it returns the same power if the input is a power of 2. Second, it
returns 0 if the input is negative or 0, while the smallest possible
integral power of 2 actually is 1 (2^0 = 1). Due to the implementation
and how it is used in a lot of places, it does not seem wise to change
such a core function however, and I decided it is better to alter the
description of the built-in.

Added a few examples/clarifications/edge-cases.

Meriipu commented:
Rebased on master and combined the recently added line https://github.com/godotengine/godot/commit/83e324d670ca05993403c5d0ad7762aa096a7978#diff-c2b842d21c6ecb3a99e6f99ba05e3dcfR1117  about complex number in C# into the note.
Why was this comment added to the gdscript document in the first place though?

aaronfranke commented:
We don't have any plans to add complex numbers to GDScript (as their use case in game development is very low), but I figured I would mention that it's supported in C#, since if someone is creating a math-heavy game then they may wish to switch to C#.

aaronfranke commented on modules/gdscript/doc_classes/@GDScript.xml:
```diff
@@ -505,6 +505,8 @@
 			</argument>
 			<description>
 				Returns [code]true[/code] if [code]a[/code] and [code]b[/code] are approximately equal to each other.
+				Here, approximately equal means that [code]a[/code] and [code]b[/code] are within a small, internal constant epsilon of each other.
```
```suggestion
				Here, approximately equal means that [code]a[/code] and [code]b[/code] are within a small internal epsilon of each other, which scales with the magnitude of the numbers.
```

Meriipu replied:
It is the smallest difference detectable that scales, rather than the internal constant, right?
`core/math/math_defs.h:#define CMP_EPSILON 0.00001`

aaronfranke commented on modules/gdscript/doc_classes/@GDScript.xml:
```diff
@@ -1093,12 +1103,15 @@
 			<argument index="1" name="to" type="float">
 			</argument>
-			<argument index="2" name="weight" type="float">
+			<argument index="2" name="x" type="float">
```
This still says "x", did you forget to run doctool?

P.S. We should add CI checks for this eventually, I'll see if I can do it today.

Meriipu replied:
Right, sorry.
I recall now that internally after clamping `p_weight` between the edges, the original code called the clamped variable `x`. 
These commits changed that internal variable to `s` (and the parameter to `p_s`).

Meriipu commented:
Is doctool supposed to update the xml-file with changes I did not make? I guess if somebody else did not run it or update the files.

I have been changing the parameters in the xml-file manually because of how few changes there were (and no new functions), so me running doctool it did not add anything that I updated this second commit.

The doctool stuff is all contained in the second of the two commits.

Meriipu commented:
leaving out the unrelated doctool updates since CI is complaining

akien-mga commented:
Thanks!

Meriipu commented:
There may be some leftover issues. It turns out that some RPCs get sent (but never accepted; they are rightfully flagged as errors on the other end) despite the master/puppet flags of the client suggesting otherwise (and there is some asymmetry here in that the server does not "wrongly" send the RPC to clients in a similar fashion). 

I assume it is a bug rather than intentional but I might be wrong. I might open a more detailed issue when I find the time.

kleonc commented:
@akien-mga Can this be cherry-picked for 3.2?

akien-mga commented:
Cherry-picked for 3.2.4.
